node/lib/util/sparse_checkout_util.js

```diff
@@ -94,3 +87,97 @@ exports.setSparseMode = co.wrap(function *(repo) {
     yield fs.writeFile(exports.getSparseCheckoutPath(repo), ".gitmodules\n");
 });
+
+/**
+ * This bit is set in the `flagsExtended` field of a `NodeGit.Index.Entry` for
+ * paths that should be skipped due to sparse checkout.
+ */
+const SKIP_WORKTREE = 1 << 14;
+
+/**
+ * Return the contents of the `.git/info/sparse-checkout` file for the
+ * specified `repo`.
+ *
+ * @param {NodeGit.Repository} repo
+ * @return {String}
+ */
+exports.readSparseCheckout = function (repo) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    const filePath = exports.getSparseCheckoutPath(repo);
+    try {
+        return fs.readFileSync(filePath, "utf8");
+    } catch (e) {
+        if ("ENOENT" !== e.code) {
+            throw e;
+        }
+        return "";
+    }
+};
+
+/**
+ * Add the specified `filename` to the set of files visible in the sparse
+ * checkout in the specified `repo`.
+ *
+ * @param {NodeGit.Repository} repo
+ * @param {String}             filename
+ */
+exports.addToSparseCheckoutFile = co.wrap(function *(repo, filename) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.isString(filename);
+
+    const sparsePath = exports.getSparseCheckoutPath(repo);
+    yield fs.appendFile(sparsePath, filename + "\n");
+});
+
+/**
+ * Remove the specified `filenames` from the set of files visible in the sparse
+ * checkout in the specified `repo`.
+ *
+ * @param {NodeGit.Repository} repo
+ * @param {String[]}           filenames
+ */
+exports.removeFromSparseCheckoutFile = function (repo, filenames) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.isArray(filenames);
+    const sparseFile = exports.readSparseCheckout(repo);
+    const toRemoveSet = new Set(filenames);
+    const newContent = sparseFile.split("\n").filter(
+                                               name => !toRemoveSet.has(name));
+    fs.writeFileSync(exports.getSparseCheckoutPath(repo),
+                     newContent.join("\n"));
+};
+
+/**
+ * Write out the specified `index` for the specified meta-repo, `repo`, set the
+ * index flags to the correct values based on the contents of
+ * `.git/info/sparse-checkout`, which libgit2 does not do.
+ *
+ * TODO: independent test
+ *
+ * @param {NodeGit.Repository} repo
+ * @param {NodeGit.Index} index
+ */
+exports.writeMetaIndex = co.wrap(function *(repo, index) {
+    assert.instanceOf(repo, NodeGit.Repository);
+    assert.instanceOf(index, NodeGit.Index);
+
+    // If we're in sparse mode, manually set bits to skip the worktree since
+    // libgit2 will not.
+
+    if (yield exports.inSparseMode(repo)) {
+        const sparseCheckout = exports.readSparseCheckout(repo);
+        const sparseSet = new Set(sparseCheckout.split("\n"));
+        const NORMAL = 0;
+        for (const e of index.entries()) {
+            if (NORMAL === NodeGit.Index.entryStage(e)) {
+                if (sparseSet.has(e.path)) {
+                    e.flagsExtended ^= SKIP_WORKTREE;
```

Comment:
We're assuming here that it's presently set and we want it unset -- but what if it wasn't previously set?  You want e.flagsExtended &= ~SKIP_WORKTREE, I think.

Reply:
Yeah, you're right.